### PR TITLE
Code - OSS Development Container

### LINKED ISSUE #339: Add dev container definition for developing VS Code
We would like to demonstrate the development of applications with a GUI component on Linux using VS Code as the example.

Current working draft is here: https://github.com/microsoft/vscode/tree/clantz/codespace

Goals:
1. Ensure this works on all platforms - macOS, Windows, Linux.
2. Ensure minimal setup - techniques like X forwarding are problematic for this reason.
3. Enable use in Codespaces over time using noVNC - this is blocked by https://github.com/MicrosoftDocs/vsonline/issues/117 currently, but we can get it in place for once it is available.

Non-goals: Provide a full desktop environment (e.g. XFCE). The goal would be to provide a way to start up VS Code and use VNC (or the web via noVNC) to see the GUI in a minimal desktop environment.

### PR DESCRIPTION
**Pull request by Chuxel**
This PR fixes resolves https://github.com/microsoft/vscode-dev-containers/issues/339.

It does the following:
1. Adds a development container for use with the VS Code Remote - Containers extension or VS Codespaces.
2. Adds a README that explains its use and limitations.
3. Bumps up a few timeouts in launch.json that I seemed to hit during testing.

The container uses the Fluxbox window manager and sets up a VNC server and no-VNC for access to the desktop environment through a web browser. Firefox is also installed by default to allow testing of APIs that launch a browser, but this can be disabled through build arguments in `devcontainer.json`.

More detail can be found in the [README file](https://github.com/microsoft/vscode/blob/clantz/dev-container/.devcontainer/README.md).

Limitations:
1. No encryption is applied to the VNC channel, but since VNC only accepts connections to localhost and forwarded over Docker's connection to the host (or via Codespace's existing encrypted channel), this should not pose a risk. Certain VNC clients may, however, mention that the channel is not encrpyted.
2. Due to MicrosoftDocs/vsonline#117, you cannot access the desktop environment from the browser when using the browser-based editor for VS Codespaces. You can, however, use VS Code to access a Codespace instead and everything works. This limitation will go away when the related issue is resolved.

### CODE REVIEW

**Comment by Chuxel** [comment]
Build failures are unrelated to this PR and are instead tied to the activation of SAML auth for Microsoft repos.

**Review by chrmarti — APPROVED** [review]
Good stuff!

File: .devcontainer/README.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+6. After the container is running, open a web browser and go to [http://localhost:6080](http://localhost:6080) or use a [VNC Viewer](https://www.realvnc.com/en/connect/download/viewer/) to connect to `localhost:5901`.
+
+Anything you start in VS Code or the integrated terminal will appear here.
+
+Next: **[Try it out!](#try-it)**
+
+## Quick start - Codespaces
+
+Note that the Codspaces browser-based editor cannot currently access the desktop environment in this containe (due to a [missing feature](https://github.com/MicrosoftDocs/vsonline/issues/117)). We recommend using Visual Studio Code from the desktop to connect instead in the near term.
```
Comment (gjsjohnmurray) [inline .devcontainer/README.md:33]: typo: containe -> container

File: .devcontainer/README.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+1. Install [Visual Studio Code Stable](https://code.visualstudio.com/) or [Insiders](https://code.visualstudio.com/insiders/) and the [Visual Studio Codespaces](https://aka.ms/vscs-ext-vscode) extension.
+
+    ![Image of VS Codespaces extension](https://microsoft.github.io/vscode-remote-release/images/codespaces-extn.png)
+
+    > Note that the Visual Studio Codespaces extension requires the Visual Studio Code distribution of Code - OSS.
+
+2. Sign in by pressing <kbd>Ctrl/Cmd</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd> and selecting **Codespaces: Sign In**. You may also need to use the **Codespaces: Create Plan** if you do not have a plan. See the [Codespaces docs](https://aka.ms/vso-docs/vscode) for details.
+
+3. Press <kbd>Ctrl/Cmd</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd> and select **Codespaes: Create New Codespace**.
```
Comment (gjsjohnmurray) [inline .devcontainer/README.md:43]: typo: Codespaes -> Codespaces

**Comment by gjsjohnmurray** [comment]
@Chuxel By following your readme I have successfully built and launched Code - OSS on Visual Studio Codespaces using your branch. It's very cool!
image

Then I exited Code - OSS (which I launched from the VS Code terminal), and tried the context menu on the noVNC window:
image

The result was:
image

**Comment by Chuxel** [comment]
@gjsjohnmurray Thanks for pointing that out. Let me look into that and see if I can repro.

**Comment by gjsjohnmurray** [comment]
Just for fun I also had a shot at launching it on GitHub Codespaces. It built OK. First time I launched it from VS Code Terminal using scripts/code.sh I got these messages in the output:
```
...
[1195:0611/145457.361738:ERROR:gl_surface_glx.cc(455)] glxQueryVersion failed
[1195:0611/145457.361800:ERROR:gl_initializer_x11.cc(148)] GLSurfaceGLX::InitializeOneOff failed.
[1195:0611/145457.367070:ERROR:viz_main_impl.cc(176)] Exiting GPU process due to errors during initialization
[1082:0611/145457.389514:WARNING:gpu_process_host.cc(1221)] The GPU process has crashed 1 time(s)
[1082:0611/145457.868786:WARNING:gpu_process_host.cc(991)] Reinitialized the GPU process after a crash. The reported initialization time was 438 ms
...
```

I aborted it (Ctrl+C), then launched again with the `--disable-gpu` option. That eliminated the GPU crash messages.

But when I tried to connect to the noVNC desktop I got this:

image

**Comment by Chuxel** [comment]
@gjsjohnmurray Yes, the GPU issue is harmless warnings and doesn't have a functional impact - it just falls back. For those on Linux, they can actually forward their GPU in.

The README mentions that Codespaces only works from VS Code due to a limitation of the port forwarding implementation:

> Note that the Codespaces browser-based editor cannot currently access the desktop environment in this container (due to a [missing feature](https://github.com/MicrosoftDocs/vsonline/issues/117)). We recommend using Visual Studio Code from the desktop to connect instead in the near term.

The Codespaces team is aware and looking into resolving it.

**Comment by gjsjohnmurray** [comment]
@Chuxel thanks for the pointer. I was able to connect my local VS Code to my GitHub Codespaces instance of your branch. The local port forwarding allowed me to get to Code - OSS on the remote desktop using noVNC. Only glitch was this:

image

**Comment by Chuxel** [comment]
>  Only glitch was this:
>
> image

@gjsjohnmurray Weird given that fall is installed by the fluxbox package in Debian. What OS are you running locally?

**Comment by gjsjohnmurray** [comment]
> 
> @gjsjohnmurray Weird given that file is installed by the fluxbox package in Debian. What OS are you running locally?

Windows 10.

**Comment by Chuxel** [comment]
@gjsjohnmurray I think I figured out what was going on with the other issues (these were specific to bind mounts on the Windows NTFS filesystem), but haven't been able to repro the fbsetbg issue.  In theory, it's possible it got resolved with one of my other edits.

**Pull request merged**